### DISCUSSION TITLE
Why Django?

### DISCUSSION THREAD

**Discussion by 3x10RaiseTo8** [vintasoftware/django-ai-assistant#163] [Q&A] [answered]
Why use Django? This is a question coming from a new, novice developer.

- Langchain and its ecosystem has embraced FastAPI and Pydantic. Maybe that's exactly why Django, because there is not much content for a group of people using a library to work with.
- The company is masters in Django so that's that.
- The project chose Django Ninja, which is similar to FastAPI. Can you share your learnings from this decision?

**Comment by fjsj** [comment, marked as answer]
Thanks for the question @3x10RaiseTo8 

You're right on all your hypotheses!

- There are several thousand Django applications out there, many of them still being maintained, new ones still being launched, and we couldn't find a trivial way to integrate Django with Langchain, so we built django-ai-assistant. There's market for something like it, so we built it.
- Our company, [Vinta Sofware](https://www.vintasoftware.com/) is a reference in Django community. We maintain one of the largest Django project templates, [django-react-boilerplate](https://github.com/vintasoftware/django-react-boilerplate)
- Django Ninja bridges the gap between traditional Django and "modern" FastAPI-like web dev. We decided to use it instead of Django REST Framework to experiment with it, and to have a simpler codebase. That said, we still want to add support to DRF in the future. Again, the goal is to facilitate integrating AI features to Django projects, including existing projects, so we must support whatever tools the Django community tends to use.

Let me know if you still have further questions.
As for FastAPI, we have ideas on how to improve Langchain support to it... stay tuned.

**Reply by 3x10RaiseTo8** [reply]
Thank you @fjsj for a good, kind answer. Django's Community seems strong and collaborative, it speaks for itself. I actively avoided working with Django for being too complicated and rigid. But the more I'm growing as a Developer, the more it is making sense. I am looking forward to making an application with Django 

And thanks for sharing your thought on FastAPI. I'm following the company page! I imagine people will mostly integrate AI as microservice, so a quick FastAPI or even Starlette app would be cool.

On another note, Langchain Inc. has been pushing LangGraph as their flagship. It is verbose, but keeps developer sanity in check. But as of now, the company has shifted focus on Cloud offerings completely neglecting self hosted solutions. The idea is to double down on LangGraph by making `langgraph-fastapi` integration. But this is too ambitious for a dev like me. Anyway, thank you for introducing me to Django Community. :smile:

**Reply by fjsj** [reply]
Thanks too! Do you know good existing langgraph + fastapi integrations?

**Reply by 3x10RaiseTo8** [reply]
Here is the list.
1. [Agent Service Toolkit](https://github.com/JoshuaC215/agent-service-toolkit): Project with most similar goal
2. [LangGraph Expense Tracker](https://github.com/jwa91/LangGraph-Expense-Tracker):
   - Dead simple structure
   - LangGraph + FastAPI + Postgres
   - Highlight: Postgres, hence can help with pgvector, a major component in RAG
   - Production-status?: need to clear more edges? dep management, updates (running langgraph 0.0.x, current 0.2) and more
3. [Conversational Agent Langchain (LangGraph)](https://github.com/mfmezger/conversational-agent-langchain):
   - Most modern codebases among all listed here.
4. [Coding-Crashkurse](https://github.com/Coding-Crashkurse) Markus - [Youtube](https://www.youtube.com/@codingcrashcourses8533): Fantastic guy with lots of microprojects. Some listed below
   - [LangGraph Fullstack Human in the loop microproject](https://github.com/Coding-Crashkurse/LangGraph-Fullstack---Human-in-the-loop)
   - [LangGraph Visualiser](https://github.com/Coding-Crashkurse/LangGraph-Visualizer): It is LangGraph+FastAPI but for LangGraph itself. Visualises graph similar to LangGraph Cloud. Use of langchain_core BaseCallbackHandler finally!!!
5. [langgraph-fastapi-streamlit-agent](https://github.com/VanJohnPK/langgraph-fastapi-streamlit-agent) (Based on agent-service-toolkit mentioned above)
6. [Finacial Chat](https://github.com/wshobson/financial-chat): Crazyyy graph, crazyyy usecase. Although uses [Langserve](https://github.com/langchain-ai/langserve) with FastAPI so not as flexible.

**Reply by fjsj** [reply]
thanks! that's super helpful.
